### PR TITLE
Add partition api and utilize sort backend

### PR DESCRIPTION
**Pull request by mfoerste4**
I added support for 
* cunumeric.partition
* cunumeric.argpartition
* ndarray.partition
* ndarray.argpartition

For now the deferred implementation utilizes sort.

This PR requires https://github.com/nv-legate/cunumeric/pull/286 in order for all tests to run.

### CODE REVIEW

**Comment by magnatelee** [comment]
@mfoerste4 please add the functions to API reference as well: https://github.com/nv-legate/cunumeric/blob/branch-22.05/docs/cunumeric/source/api/sorting.rst

To build the doc yourself, run `make html` in `docs/cunumeric`. There shouldn't be any error when you're doing it.

File: tests/partition.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    shape = a.shape
+    volume = 1
+    for i in range(a.ndim):
+        volume *= shape[i]
+
+    # partition axes
+    for i in range(a.ndim):
+        kth = math.floor(shape[i] / 2)
+        print("partition axis " + str(i))
```
Comment (magnatelee) [inline tests/partition.py:80]: Use an f-string.

File: tests/partition.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        assert_partition(
+            num.partition(a_num, kth=kth, axis=i).__array__(), kth, i
+        )
+
+    # flatten
+    print("partition flattened")
+    kth = math.floor(volume / 2)
```
Comment (magnatelee) [inline tests/partition.py:87]: this could be `kth = volume // 2`

File: tests/partition.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    assert_partition(
+        num.partition(a_num, kth=kth, axis=None).__array__(), kth, 0
+    )
+
+    # in-place partition
+    kth = math.floor(shape[a.ndim - 1] / 2)
```
Comment (magnatelee) [inline tests/partition.py:93]: same here. `math.floor` is unnecessary.

File: tests/partition.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    copy_a_num = a_num.copy()
+    copy_a_num.partition(kth)
+    assert_partition(copy_a_num.__array__(), kth, a.ndim - 1)
+
+    # argpartition
+    for i in range(a.ndim):
+        kth = math.floor(shape[i] / 2)
```
Comment (magnatelee) [inline tests/partition.py:100]: same here

File: tests/partition.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        assert_argpartition(
+            num.argpartition(a_num, kth, axis=i).__array__(),
+            a_num.__array__(),
+            kth,
+            i,
+        )
+
+    # flatten
+    print("argpartition flattened")
+    kth = math.floor(volume / 2)
```
Comment (magnatelee) [inline tests/partition.py:111]: same here

File: tests/partition.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    assert_argpartition(
+        num.argpartition(a_num, kth, axis=None).__array__(),
+        a_num.flatten().__array__(),
+        kth,
+        0,
+    )
+
+    # nd.argpartition -- no change to array
+    kth = math.floor(shape[a.ndim - 1] / 2)
```
Comment (magnatelee) [inline tests/partition.py:120]: same here

File: tests/partition.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    a_num = num.array(a)
+
+    shape = a.shape
+    volume = 1
+    for i in range(a.ndim):
+        volume *= shape[i]
+
+    # partition axes
+    for i in range(a.ndim):
+        kth = math.floor(shape[i] / 2)
```
Comment (magnatelee) [inline tests/partition.py:79]: this looks like the last one

File: tests/partition.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        print("partition axis " + str(i))
+        assert_partition(
+            num.partition(a_num, kth=kth, axis=i).__array__(), kth, i
```
Comment (magnatelee) [inline tests/partition.py:82]: if you will call `__array__` for all these outputs, can't you just do it inside `assert_argpartition`?

**Comment by mfoerste4** [comment]
> @mfoerste4 please add the functions to API reference as well: https://github.com/nv-legate/cunumeric/blob/branch-22.05/docs/cunumeric/source/api/sorting.rst
> 
> To build the doc yourself, run `make html` in `docs/cunumeric`. There shouldn't be any error when you're doing it.

I updated the sorting.rst as well as (_)ndarray.rst files. However, the build showed errors regarding 'window' and 'bartlett' which seem unrelated.

**Comment by magnatelee** [comment]
> I updated the sorting.rst as well as (_)ndarray.rst files. However, the build showed errors regarding 'window' and 'bartlett' which seem unrelated.

There was a doc change that hadn't been pushed to the repository. The problem should be fixed once you pull branch-22.05.

**Comment by mfoerste4** [comment]
> There was a doc change that hadn't been pushed to the repository. The problem should be fixed once you pull branch-22.05.
The merge fixed it.

**Review by magnatelee — APPROVED** [review]

**Pull request merged**